modules/dag-l1/src/main/scala/org/tessellation/dag/l1/storage/TransactionStorage.scala

```diff
@@ -0,0 +1,111 @@
+package org.tessellation.dag.l1.storage
+
+import cats.data.{NonEmptyList, NonEmptySet}
+import cats.effect.{Ref, Sync}
+import cats.syntax.applicative._
+import cats.syntax.flatMap._
+import cats.syntax.functor._
+import cats.syntax.option._
+
+import scala.collection.immutable.SortedSet
+
+import org.tessellation.dag.transaction.filter.Consecutive
+import org.tessellation.ext.collection.MapRefUtils._
+import org.tessellation.schema.address.Address
+import org.tessellation.schema.transaction.{Transaction, TransactionReference}
+import org.tessellation.security.Hashed
+import org.tessellation.security.signature.Signed
+
+import io.chrisdavenport.mapref.MapRef
+
+trait TransactionStorage[F[_]] {
+  def isParentAccepted(tx: Transaction): F[Boolean]
+
+  def getLastAcceptedTransactionRef(address: Address): F[TransactionReference]
+
+  def acceptTransaction(hashedTx: Hashed[Transaction]): F[Unit]
+
+  def putTransactions(transactions: Set[Signed[Transaction]]): F[Unit]
+
+  def pullTransactions(): F[Option[NonEmptyList[Signed[Transaction]]]]
+}
+
+object TransactionStorage {
+
+  def make[F[_]: Sync]: F[TransactionStorage[F]] =
+    for {
+      lastAccepted <- MapRef.ofConcurrentHashMap[F, Address, TransactionReference]()
+      waitingTransactions <- Ref.of[F, Map[Address, NonEmptySet[Signed[Transaction]]]](Map.empty)
+      transactionStorage = make(lastAccepted, waitingTransactions)
+    } yield transactionStorage
+
+  def make[F[_]: Sync](
+    lastAccepted: MapRef[F, Address, Option[TransactionReference]],
+    waitingTransactions: Ref[F, Map[Address, NonEmptySet[Signed[Transaction]]]]
+  ): TransactionStorage[F] =
+    new TransactionStorage[F] {
+
+      def isParentAccepted(transaction: Transaction): F[Boolean] =
+        (transaction.parent == TransactionReference.empty)
+          .pure[F]
+          .ifM(
+            true.pure[F],
+            lastAccepted(transaction.source).get.map(_.contains(transaction.parent))
+          )
```

Comment:
```scala
(transaction.parent == TransactionReference.empty)
  .guard[Option]
  .fold(true.pure[F]) { _ =>
     lastAccepted(transaction.source).get.map(_.contains(transaction.parent)))
   }
```

Reply:
Needed to negate the first condition. `true` was passing as `Some`